### ISSUE TITLE
Swarm service not registered in internal DNS

### ISSUE THREAD

**Issue by kostrzewa9ld** [moby/moby#35508]
I have a docker swarm setup on 4 manager nodes (not 3 by accident). At some point one of the nodes got out of space on root partition. After a while (around 2 hours) swarm run leader election and all the services were restarted. However, restart of the services didn't go as expected. For example, some services stabilized after around 24 hours(!) and some didn't stabilize till now and in a result they are no longer resolvable by its name (dns query go to our external DNS).

**Steps to reproduce the issue:**
I don't have exact reproduction steps, but the services restart might have something to do with one of the nodes getting out of space and election of a new leader.

**Describe the results you received:**
Some services are not resolvable by their name.

**Describe the results you expected:**
When service is restarted it should be properly registered in swarm.

**Additional information you deem important (e.g. issue happens only occasionally):**
Services were deployed from docker compose files using `docker stack deploy` so the issue might be to some extent related to #34904.

**Logs**
Logs related to the service that didn't stabilize at all (i.e. wasn't registered in swarm's internal DNS).
```
Nov 12 22:05:53 node-10-9-4-103 dockerd: time="2017-11-12T22:05:53.729702654+01:00" level=debug msg="Name To resolve: sa_openldap_replica."
Nov 12 22:05:53 node-10-9-4-103 dockerd: time="2017-11-12T22:05:53.729821797+01:00" level=debug msg="Lookup for sa_openldap_replica.: IP [10.0.0.22]"
Nov 12 22:05:53 node-10-9-4-103 dockerd: time="2017-11-12T22:05:53.729702787+01:00" level=debug msg="Name To resolve: sa_openldap_replica."
Nov 12 22:05:53 node-10-9-4-103 dockerd: time="2017-11-12T22:05:53.729920814+01:00" level=debug msg="Lookup name sa_openldap_replica. present without IPv6 address"
Nov 12 22:06:00 node-10-9-4-103 dockerd: time="2017-11-12T22:06:00.680899703+01:00" level=debug msg="Name To resolve: sa_openldap_replica."
Nov 12 22:06:00 node-10-9-4-103 dockerd: time="2017-11-12T22:06:00.681042598+01:00" level=debug msg="Lookup name sa_openldap_replica. present without IPv6 address"
Nov 12 22:06:00 node-10-9-4-103 dockerd: time="2017-11-12T22:06:00.684198454+01:00" level=debug msg="Name To resolve: sa_openldap_replica."
Nov 12 22:06:00 node-10-9-4-103 dockerd: time="2017-11-12T22:06:00.684247545+01:00" level=debug msg="Lookup for sa_openldap_replica.: IP [10.0.0.22]"
Nov 12 22:06:11 node-10-9-4-103 dockerd: time="2017-11-12T22:06:11.735412966+01:00" level=debug msg="deleteServiceInfoFromCluster from sbLeave START for sa_openldap_replica 426a580e53fd721d410d3a2b087bf8cb998b9ecb6a0cd8da5c8b00f87cb93a90"
Nov 12 22:06:11 node-10-9-4-103 dockerd: time="2017-11-12T22:06:11.735571015+01:00" level=debug msg="rmServiceBinding from deleteServiceInfoFromCluster START for sa_openldap_replica 426a580e53fd721d410d3a2b087bf8cb998b9ecb6a0cd8da5c8b00f87
cb93a90 p:0xc4219f5480 nid:ai6hs0uuezm9hlqpi44aw0w22 sKey:{ljrvybzqa6b5z2m67643einto } deleteSvc:true"
Nov 12 22:06:12 node-10-9-4-103 dockerd: time="2017-11-12T22:06:12.000830296+01:00" level=debug msg="deleteEndpointNameResolution 426a580e53fd721d410d3a2b087bf8cb998b9ecb6a0cd8da5c8b00f87cb93a90 sa_openldap_replica rm_service:true suppress
:false sAliases:[openldap_replica] tAliases:[1951c4f0d3ce]"
Nov 12 22:06:12 node-10-9-4-103 dockerd: time="2017-11-12T22:06:12.000902728+01:00" level=debug msg="delContainerNameResolution 426a580e53fd721d410d3a2b087bf8cb998b9ecb6a0cd8da5c8b00f87cb93a90 sa_openldap_replica.1.kq869yybtx4bq9437m5cqh57
f"
Nov 12 22:06:12 node-10-9-4-103 dockerd: time="2017-11-12T22:06:12.000931567+01:00" level=debug msg="426a580e53fd721d410d3a2b087bf8cb998b9ecb6a0cd8da5c8b00f87cb93a90 (ai6hs0u).deleteSvcRecords(sa_openldap_replica.1.kq869yybtx4bq9437m5cqh57
f, 10.0.0.23, <nil>, true) rmServiceBinding sid:426a580e53fd721d410d3a2b087bf8cb998b9ecb6a0cd8da5c8b00f87cb93a90 "
Nov 12 22:06:12 node-10-9-4-103 dockerd: time="2017-11-12T22:06:12.001064630+01:00" level=debug msg="426a580e53fd721d410d3a2b087bf8cb998b9ecb6a0cd8da5c8b00f87cb93a90 (ai6hs0u).deleteSvcRecords(tasks.sa_openldap_replica, 10.0.0.23, <nil>, f
alse) rmServiceBinding sid:ljrvybzqa6b5z2m67643einto "
Nov 12 22:06:12 node-10-9-4-103 dockerd: time="2017-11-12T22:06:12.001128031+01:00" level=debug msg="426a580e53fd721d410d3a2b087bf8cb998b9ecb6a0cd8da5c8b00f87cb93a90 (ai6hs0u).deleteSvcRecords(sa_openldap_replica, 10.0.0.22, <nil>, false) 
rmServiceBinding sid:ljrvybzqa6b5z2m67643einto "
Nov 12 22:06:12 node-10-9-4-103 dockerd: time="2017-11-12T22:06:12.001178124+01:00" level=debug msg="rmServiceBinding from deleteServiceInfoFromCluster END for sa_openldap_replica 426a580e53fd721d410d3a2b087bf8cb998b9ecb6a0cd8da5c8b00f87cb
93a90"
Nov 12 22:06:12 node-10-9-4-103 dockerd: time="2017-11-12T22:06:12.001199826+01:00" level=debug msg="deleteServiceInfoFromCluster from sbLeave END for sa_openldap_replica 426a580e53fd721d410d3a2b087bf8cb998b9ecb6a0cd8da5c8b00f87cb93a90"
Nov 12 22:06:17 node-10-9-4-103 dockerd: time="2017-11-12T22:06:17.878801373+01:00" level=debug msg="Releasing addresses for endpoint sa_openldap_replica.1.kq869yybtx4bq9437m5cqh57f's interface on network sa"
Nov 12 22:07:33 node-10-9-4-103 dockerd: time="2017-11-12T22:07:33.407972442+01:00" level=debug msg="copying image data from 3d18bae8e135b131302e67e6b723664c71e69a49a403516800ac298bd4e7dd2b:/var/lib/ldap-single, to sa_openldap_replica"
Nov 12 22:07:33 node-10-9-4-103 dockerd: time="2017-11-12T22:07:33.526664394+01:00" level=debug msg="Creating service config in agent for t = &Task{ID:xxnkynuwa0xroymvzjcqsr2wc,Meta:Meta{Version:Version{Index:596,},CreatedAt:2017-11-12T21:
05:53.981511233Z,UpdatedAt:2017-11-12T21:05:54.137337895Z,},Spec:TaskSpec{Runtime:&TaskSpec_Container{Container:&ContainerSpec{Image:ramencloud/fusiondirectory-openldap:0.3.0@sha256:141a737f228a67aee03b8ac3fb2e3d7e2fb2e8596e9cdbc0ee60a26ab
5ea8d25,Labels:map[string]string{com.docker.stack.namespace: sa,},Command:[],Args:[],Env:[<CAN'T_SHARE_THESE>],Dir:,User:,Mounts:[{VOLUME sa_openld
ap_replica /var/lib/ldap-single false nil Mount_VolumeOptions{NoCopy:false,Labels:map[string]string{com.docker.stack.namespace: sa,},DriverConfig:&Driver{Name:local,Options:map[string]string{},},} nil}],StopGracePeriod:(duration: nil googl
e_protobuf1.Duration),PullOptions:nil,Groups:[],Secrets:[],TTY:false,Hostname:,DNSConfig:nil,Healthcheck:&HealthConfig{Test:[CMD-SHELL ldapsearch -Q -Y EXTERNAL -H ldapi:/// -b \"cn=config\" \"cn=config\"],Interval:10s,Timeout:5s,Retries:3
0,StartPeriod:0s,},Hosts:[],OpenStdin:false,ReadOnly:false,StopSignal:,Configs:[],Privileges:&Privileges{CredentialSpec:nil,SELinuxContext:nil,},},},Resources:&ResourceRequirements{Limits:nil,Reservations:nil,},Restart:nil,Placement:&Place
ment{Constraints:[node.labels.openldap.role == replica],Preferences:[],Platforms:[&Platform{Architecture:amd64,OS:linux,}],},LogDriver:nil,Networks:[&NetworkAttachmentConfig{Target:ai6hs0uuezm9hlqpi44aw0w22,Aliases:[openldap_replica],Addre
sses:[],DriverAttachmentOpts:map[string]string{},}],ForceUpdate:0,},ServiceID:ljrvybzqa6b5z2m67643einto,Slot:1,NodeID:jvxg46j8bcwa8o5qa96psner9,Annotations:Annotations{Name:,Lab
Nov 12 22:07:33 node-10-9-4-103 dockerd: els:map[string]string{},Indices:[],},ServiceAnnotations:Annotations{Name:sa_openldap_replica,Labels:map[string]string{com.docker.stack.image: ramencloud/fusiondirectory-openldap:0.3.0,com.docker.sta
ck.namespace: sa,},Indices:[],},Status:TaskStatus{Timestamp:2017-11-12T21:05:54.136957131Z,State:ASSIGNED,Message:scheduler assigned task to node,Err:,RuntimeStatus:<nil>,PortStatus:nil,},DesiredState:READY,Networks:[&NetworkAttachment{Net
work:&Network{ID:ai6hs0uuezm9hlqpi44aw0w22,Meta:Meta{Version:Version{Index:582,},CreatedAt:2017-10-30T15:01:11.480607346Z,UpdatedAt:2017-11-12T21:05:45.11102014Z,},Spec:NetworkSpec{Annotations:Annotations{Name:sa,Labels:map[string]string{}
,Indices:[],},DriverConfig:&Driver{Name:overlay,Options:map[string]string{},},Ipv6Enabled:false,Internal:false,IPAM:nil,Attachable:true,Ingress:false,ConfigFrom:,},DriverState:&Driver{Name:overlay,Options:map[string]string{com.docker.netwo
rk.driver.overlay.vxlanid_list: 4097,},},IPAM:&IPAMOptions{Driver:&Driver{Name:default,Options:map[string]string{},},Configs:[&IPAMConfig{Family:IPV4,Subnet:10.0.0.0/24,Range:,Gateway:10.0.0.1,Reserved:map[string]string{},}],},},Addresses:
[10.0.0.35/24],Aliases:[openldap_replica],DriverAttachmentOpts:map[string]string{},}],Endpoint:&Endpoint{Spec:&EndpointSpec{Mode:VIP,Ports:[],},Ports:[],VirtualIPs:[&Endpoint_VirtualIP{NetworkID:ai6hs0uuezm9hlqpi44aw0w22,Addr:10.0.0.22/24,
}],},LogDriver:nil,SpecVersion:&Version{Index:0,},}"
Nov 12 22:07:33 node-10-9-4-103 dockerd: time="2017-11-12T22:07:33.594981939+01:00" level=debug msg="Assigning addresses for endpoint sa_openldap_replica.1.xxnkynuwa0xroymvzjcqsr2wc's interface on network sa"
Nov 12 22:07:33 node-10-9-4-103 dockerd: time="2017-11-12T22:07:33.597121564+01:00" level=debug msg="Assigning addresses for endpoint sa_openldap_replica.1.xxnkynuwa0xroymvzjcqsr2wc's interface on network sa"
Nov 12 22:10:28 node-10-9-4-103 dockerd: time="2017-11-12T22:10:28.709341955+01:00" level=debug msg="Name To resolve: sa_openldap_replica."
Nov 12 22:10:28 node-10-9-4-103 dockerd: time="2017-11-12T22:10:28.709474549+01:00" level=debug msg="Name To resolve: sa_openldap_replica."
Nov 12 22:10:28 node-10-9-4-103 dockerd: time="2017-11-12T22:10:28.726655795+01:00" level=debug msg="Query sa_openldap_replica.[1] from 172.18.0.5:59602, forwarding to udp:10.9.4.4"
Nov 12 22:10:28 node-10-9-4-103 dockerd: time="2017-11-12T22:10:28.732668381+01:00" level=debug msg="Query sa_openldap_replica.[28] from 172.18.0.5:40307, forwarding to udp:10.9.4.4"
Nov 12 22:10:28 node-10-9-4-103 dockerd: time="2017-11-12T22:10:28.774710890+01:00" level=debug msg="Name To resolve: sa_openldap_replica.bdcluster."
Nov 12 22:10:28 node-10-9-4-103 dockerd: time="2017-11-12T22:10:28.774792259+01:00" level=debug msg="Name To resolve: sa_openldap_replica.bdcluster."
```
The last messages - "Query" and "Name To resolve" - keep repeating until now.

Logs related to service that was restarted after around 24hours:
```
Nov 12 22:06:17 node-10-9-4-103 dockerd: time="2017-11-12T22:06:17.398920379+01:00" level=debug msg="deleteServiceInfoFromCluster from sbLeave START for sa_postgres_replica b2cfde017504b180d1bc02a1d90725451b48efaaa3185eec4bc70a3ece61c582"
Nov 12 22:06:17 node-10-9-4-103 dockerd: time="2017-11-12T22:06:17.399015314+01:00" level=debug msg="rmServiceBinding from deleteServiceInfoFromCluster START for sa_postgres_replica b2cfde017504b180d1bc02a1d90725451b48efaaa3185eec4bc70a3ece61c582 p:0xc420389780 nid:ai6hs0uuezm9hlqpi44aw0w22 sKey:{lgczmoowxriiseuxksl7fibzw } deleteSvc:true"
Nov 12 22:06:19 node-10-9-4-103 dockerd: time="2017-11-12T22:06:19.538822724+01:00" level=debug msg="deleteEndpointNameResolution b2cfde017504b180d1bc02a1d90725451b48efaaa3185eec4bc70a3ece61c582 sa_postgres_replica rm_service:true suppress:false sAliases:[postgres_replica] tAliases:[f959938f9f67]"
Nov 12 22:06:19 node-10-9-4-103 dockerd: time="2017-11-12T22:06:19.538877237+01:00" level=debug msg="delContainerNameResolution b2cfde017504b180d1bc02a1d90725451b48efaaa3185eec4bc70a3ece61c582 sa_postgres_replica.1.xi6oi1yy6kfp5nkqdfit3ypzd"
Nov 12 22:06:19 node-10-9-4-103 dockerd: time="2017-11-12T22:06:19.538904963+01:00" level=debug msg="b2cfde017504b180d1bc02a1d90725451b48efaaa3185eec4bc70a3ece61c582 (ai6hs0u).deleteSvcRecords(sa_postgres_replica.1.xi6oi1yy6kfp5nkqdfit3ypzd, 10.0.0.15, <nil>, true) rmServiceBinding sid:b2cfde017504b180d1bc02a1d90725451b48efaaa3185eec4bc70a3ece61c582 "
Nov 12 22:06:19 node-10-9-4-103 dockerd: time="2017-11-12T22:06:19.539002002+01:00" level=debug msg="b2cfde017504b180d1bc02a1d90725451b48efaaa3185eec4bc70a3ece61c582 (ai6hs0u).deleteSvcRecords(tasks.sa_postgres_replica, 10.0.0.15, <nil>, false) rmServiceBinding sid:lgczmoowxriiseuxksl7fibzw "
Nov 12 22:06:19 node-10-9-4-103 dockerd: time="2017-11-12T22:06:19.539090778+01:00" level=debug msg="b2cfde017504b180d1bc02a1d90725451b48efaaa3185eec4bc70a3ece61c582 (ai6hs0u).deleteSvcRecords(sa_postgres_replica, 10.0.0.18, <nil>, false) rmServiceBinding sid:lgczmoowxriiseuxksl7fibzw "
Nov 12 22:06:19 node-10-9-4-103 dockerd: time="2017-11-12T22:06:19.539148875+01:00" level=debug msg="rmServiceBinding from deleteServiceInfoFromCluster END for sa_postgres_replica b2cfde017504b180d1bc02a1d90725451b48efaaa3185eec4bc70a3ece61c582"
Nov 12 22:06:19 node-10-9-4-103 dockerd: time="2017-11-12T22:06:19.539184113+01:00" level=debug msg="deleteServiceInfoFromCluster from sbLeave END for sa_postgres_replica b2cfde017504b180d1bc02a1d90725451b48efaaa3185eec4bc70a3ece61c582"
Nov 12 22:06:49 node-10-9-4-103 dockerd: time="2017-11-12T22:06:49.609746534+01:00" level=debug msg="Releasing addresses for endpoint sa_postgres_replica.1.xi6oi1yy6kfp5nkqdfit3ypzd's interface on network sa"
Nov 12 22:07:22 node-10-9-4-103 dockerd: time="2017-11-12T22:07:22.540628448+01:00" level=debug msg="copying image data from 09635baa4d866d550c70c86ee1299b3abcfe93f0aa1f716040c0915711254437:/pgdata, to sa_postgres_replica"
Nov 12 22:07:32 node-10-9-4-103 dockerd: time="2017-11-12T22:07:32.893196027+01:00" level=debug msg="Creating service config in agent for t = &Task{ID:ho1w9rwm6792furoa7y8fzrwk,Meta:Meta{Version:Version{Index:596,},CreatedAt:2017-11-12T21:05:53.983038323Z,UpdatedAt:2017-11-12T21:05:54.137140678Z,},Spec:TaskSpec{Runtime:&TaskSpec_Container{Container:&ContainerSpec{Image:ramencloud/crunchy-postgres:centos7-9.6-1.5.1@sha256:073e420d5d8c41fb17487464adc5f1890a2997062c75f84ca81d5573449b86f7,Labels:map[string]string{com.docker.stack.namespace: sa,},Command:[],Args:[],Env:[<CAN'T_SHARE_THESE>],Dir:,User:,Mounts:[{VOLUME sa_postgres_replica /pgdata false nil Mount_VolumeOptions{NoCopy:false,Labels:map[string]string{com.docker.stack.namespace: sa,},DriverConfig:&Driver{Name:local,Options:map[string]string{},},} nil}],StopGracePeriod:(duration: nil google_protobuf1.Duration),PullOptions:nil,Groups:[],Secrets:[],TTY:false,Hostname:,DNSConfig:nil,Healthcheck:&HealthConfig{Test:[CMD-SHELL /usr/pgsql-9.6/bin/pg_isready --host localhost],Interval:10s,Timeout:5s,Retries:30,StartPeriod:0s,},Hosts:[],OpenStdin:false,ReadOnly:false,StopSignal:,Configs:[],Privileges:&Privileges{CredentialSpec:nil,SELinuxContext:nil,},},},Resources:&ResourceRequirements{Limits:nil,Reservations:nil,},Restart:nil,Placement:&Placement{Constraints:[node.labels.postgres.role == replica],Preferences:[],Platforms:[&Platform{Architecture:amd64,OS:linux,}],},LogDriver:nil,Networks:[&NetworkAttachmentConfig{Target:ai6hs0uuezm9hlqpi44aw0w22,Aliases:[postgres_replica],Addresses:[],DriverAttachmentOpts:map[string]string{},}],ForceUpdate:0,},ServiceID:lgczmoowxriiseuxksl7fibzw,Slot:1,NodeID:jvxg46j8bcwa8o5qa96psner9,Annotations:Annotations{Name:,Labels:map[string]string{},Indices:[],},ServiceAnnotations:Annotations{Name:sa_postgres_replica,Labels:map[string]string{com.docke
Nov 12 22:07:33 node-10-9-4-103 dockerd: time="2017-11-12T22:07:33.350005291+01:00" level=debug msg="Assigning addresses for endpoint sa_postgres_replica.1.ho1w9rwm6792furoa7y8fzrwk's interface on network sa"
Nov 12 22:07:33 node-10-9-4-103 dockerd: time="2017-11-12T22:07:33.352217431+01:00" level=debug msg="Assigning addresses for endpoint sa_postgres_replica.1.ho1w9rwm6792furoa7y8fzrwk's interface on network sa"
Nov 12 22:09:48 node-10-9-4-103 dockerd: time="2017-11-12T22:09:48.340170229+01:00" level=debug msg="Name To resolve: sa_postgres."
Nov 12 22:09:48 node-10-9-4-103 dockerd: time="2017-11-12T22:09:48.340336962+01:00" level=debug msg="Lookup name sa_postgres. present without IPv6 address"
Nov 12 22:09:48 node-10-9-4-103 dockerd: time="2017-11-12T22:09:48.340458050+01:00" level=debug msg="Name To resolve: sa_postgres."
Nov 12 22:09:48 node-10-9-4-103 dockerd: time="2017-11-12T22:09:48.340564002+01:00" level=debug msg="Lookup for sa_postgres.: IP [10.0.0.16]"
Nov 12 22:10:00 node-10-9-4-103 dockerd: time="2017-11-12T22:10:00.653118678+01:00" level=debug msg="handleEpTableEvent DEL %!s(bool=false) R:52ef29d102a59bdec28785368a7d0d8a78f2b962a8a16d5bd7223466e87b55a8%!(EXTRA libnetwork.EndpointRecord={sa_postgres.1.hwoulyjm32wb3cngpk0dhk1h2 sa_postgres 5zdfyhotm7si0tlpvwwjjrys9 10.0.0.16 10.0.0.29 [] [postgres] [6f0304e02db9]})"
Nov 12 22:10:00 node-10-9-4-103 dockerd: time="2017-11-12T22:10:00.653215576+01:00" level=debug msg="rmServiceBinding from handleEpTableEvent START for sa_postgres 52ef29d102a59bdec28785368a7d0d8a78f2b962a8a16d5bd7223466e87b55a8 p:0xc420c92000 nid:ai6hs0uuezm9hlqpi44aw0w22 sKey:{5zdfyhotm7si0tlpvwwjjrys9 } deleteSvc:true"
Nov 12 22:10:00 node-10-9-4-103 dockerd: time="2017-11-12T22:10:00.862800775+01:00" level=debug msg="deleteEndpointNameResolution 52ef29d102a59bdec28785368a7d0d8a78f2b962a8a16d5bd7223466e87b55a8 sa_postgres rm_service:true suppress:false sAliases:[postgres] tAliases:[6f0304e02db9]"
Nov 12 22:10:00 node-10-9-4-103 dockerd: time="2017-11-12T22:10:00.862854961+01:00" level=debug msg="delContainerNameResolution 52ef29d102a59bdec28785368a7d0d8a78f2b962a8a16d5bd7223466e87b55a8 sa_postgres.1.hwoulyjm32wb3cngpk0dhk1h2"
Nov 12 22:10:00 node-10-9-4-103 dockerd: time="2017-11-12T22:10:00.862881831+01:00" level=debug msg="52ef29d102a59bdec28785368a7d0d8a78f2b962a8a16d5bd7223466e87b55a8 (ai6hs0u).deleteSvcRecords(sa_postgres.1.hwoulyjm32wb3cngpk0dhk1h2, 10.0.0.29, <nil>, true) rmServiceBinding sid:52ef29d102a59bdec28785368a7d0d8a78f2b962a8a16d5bd7223466e87b55a8 "
Nov 12 22:10:00 node-10-9-4-103 dockerd: time="2017-11-12T22:10:00.862972641+01:00" level=debug msg="52ef29d102a59bdec28785368a7d0d8a78f2b962a8a16d5bd7223466e87b55a8 (ai6hs0u).deleteSvcRecords(tasks.sa_postgres, 10.0.0.29, <nil>, false) rmServiceBinding sid:5zdfyhotm7si0tlpvwwjjrys9 "
Nov 12 22:10:00 node-10-9-4-103 dockerd: time="2017-11-12T22:10:00.863039747+01:00" level=debug msg="52ef29d102a59bdec28785368a7d0d8a78f2b962a8a16d5bd7223466e87b55a8 (ai6hs0u).deleteSvcRecords(sa_postgres, 10.0.0.16, <nil>, false) rmServiceBinding sid:5zdfyhotm7si0tlpvwwjjrys9 "
Nov 12 22:10:00 node-10-9-4-103 dockerd: time="2017-11-12T22:10:00.863103500+01:00" level=debug msg="rmServiceBinding from handleEpTableEvent END for sa_postgres 52ef29d102a59bdec28785368a7d0d8a78f2b962a8a16d5bd7223466e87b55a8"
Nov 12 22:10:02 node-10-9-4-103 dockerd: time="2017-11-12T22:10:02.263681274+01:00" level=debug msg="Name To resolve: sa_postgres."
Nov 12 22:10:02 node-10-9-4-103 dockerd: time="2017-11-12T22:10:02.263882687+01:00" level=debug msg="Name To resolve: sa_postgres."
Nov 12 22:10:02 node-10-9-4-103 dockerd: time="2017-11-12T22:10:02.277630397+01:00" level=debug msg="Query sa_postgres.[28] from 172.18.0.4:55359, forwarding to udp:10.9.4.4"
Nov 12 22:10:02 node-10-9-4-103 dockerd: time="2017-11-12T22:10:02.277887855+01:00" level=debug msg="Query sa_postgres.[1] from 172.18.0.4:34120, forwarding to udp:10.9.4.4"
...
Nov 13 21:52:50 node-10-9-4-103 dockerd: time="2017-11-13T21:52:50.349596635+01:00" level=debug msg="Query sa_postgres.bdcluster.[1] from 172.18.0.4:39387, forwarding to udp:10.9.4.4"
Nov 13 21:52:50 node-10-9-4-103 dockerd: time="2017-11-13T21:52:50.349946529+01:00" level=debug msg="Query sa_postgres.bdcluster.[28] from 172.18.0.4:35307, forwarding to udp:10.9.4.4"
Nov 13 21:52:51 node-10-9-4-103 dockerd: time="2017-11-13T21:52:51.452864532+01:00" level=debug msg="handleEpTableEvent ADD %!s(bool=true) R:74d31fc4bbaeca6627dbd4b0037b34dc4635b037ff929b2c60c82c5c9780b536%!(EXTRA libnetwork.EndpointRecord={sa_postgres.1.b7u81fw6d3fykjy48p72zuec3 sa_postgres 5zdfyhotm7si0tlpvwwjjrys9 10.0.0.16 10.0.0.23 [] [postgres] [eaa9a1981266]})"
Nov 13 21:52:51 node-10-9-4-103 dockerd: time="2017-11-13T21:52:51.452984154+01:00" level=debug msg="addServiceBinding from handleEpTableEvent START for sa_postgres 74d31fc4bbaeca6627dbd4b0037b34dc4635b037ff929b2c60c82c5c9780b536 p:0xc4226c3480 nid:ai6hs0uuezm9hlqpi44aw0w22 skey:{5zdfyhotm7si0tlpvwwjjrys9 }"
Nov 13 21:52:51 node-10-9-4-103 dockerd: time="2017-11-13T21:52:51.877031972+01:00" level=debug msg="addEndpointNameResolution 74d31fc4bbaeca6627dbd4b0037b34dc4635b037ff929b2c60c82c5c9780b536 sa_postgres add_service:true sAliases:[postgres] tAliases:[eaa9a1981266]"
Nov 13 21:52:51 node-10-9-4-103 dockerd: time="2017-11-13T21:52:51.877122380+01:00" level=debug msg="addContainerNameResolution 74d31fc4bbaeca6627dbd4b0037b34dc4635b037ff929b2c60c82c5c9780b536 sa_postgres.1.b7u81fw6d3fykjy48p72zuec3"
Nov 13 21:52:51 node-10-9-4-103 dockerd: time="2017-11-13T21:52:51.877173173+01:00" level=debug msg="74d31fc4bbaeca6627dbd4b0037b34dc4635b037ff929b2c60c82c5c9780b536 (ai6hs0u).addSvcRecords(sa_postgres.1.b7u81fw6d3fykjy48p72zuec3, 10.0.0.23, <nil>, true) addServiceBinding sid:74d31fc4bbaeca6627dbd4b0037b34dc4635b037ff929b2c60c82c5c9780b536"
Nov 13 21:52:51 node-10-9-4-103 dockerd: time="2017-11-13T21:52:51.877292066+01:00" level=debug msg="74d31fc4bbaeca6627dbd4b0037b34dc4635b037ff929b2c60c82c5c9780b536 (ai6hs0u).addSvcRecords(tasks.sa_postgres, 10.0.0.23, <nil>, false) addServiceBinding sid:5zdfyhotm7si0tlpvwwjjrys9"
Nov 13 21:52:51 node-10-9-4-103 dockerd: time="2017-11-13T21:52:51.877376445+01:00" level=debug msg="74d31fc4bbaeca6627dbd4b0037b34dc4635b037ff929b2c60c82c5c9780b536 (ai6hs0u).addSvcRecords(sa_postgres, 10.0.0.16, <nil>, false) addServiceBinding sid:5zdfyhotm7si0tlpvwwjjrys9"
Nov 13 21:52:51 node-10-9-4-103 dockerd: time="2017-11-13T21:52:51.877441389+01:00" level=debug msg="addServiceBinding from handleEpTableEvent END for sa_postgres 74d31fc4bbaeca6627dbd4b0037b34dc4635b037ff929b2c60c82c5c9780b536"
Nov 13 21:52:55 node-10-9-4-103 dockerd: time="2017-11-13T21:52:55.308085731+01:00" level=debug msg="Name To resolve: sa_postgres."
Nov 13 21:52:55 node-10-9-4-103 dockerd: time="2017-11-13T21:52:55.308225796+01:00" level=debug msg="Lookup for sa_postgres.: IP [10.0.0.16]"
Nov 13 21:52:55 node-10-9-4-103 dockerd: time="2017-11-13T21:52:55.308086099+01:00" level=debug msg="Name To resolve: sa_postgres."
Nov 13 21:52:55 node-10-9-4-103 dockerd: time="2017-11-13T21:52:55.308300136+01:00" level=debug msg="Lookup name sa_postgres. present without IPv6 address"
```

**Output of `docker version`:**

```
Client:
 Version:      17.06.2-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   cec0b72
 Built:        Tue Sep  5 19:59:06 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.2-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   cec0b72
 Built:        Tue Sep  5 20:00:25 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 15
 Running: 6
 Paused: 0
 Stopped: 9
Images: 14
Server Version: 17.06.2-ce
Storage Driver: devicemapper
 Pool Name: vg00-docker--pool
 Pool Blocksize: 524.3kB
 Base Device Size: 10.74GB
 Backing Filesystem: xfs
 Data file: 
 Metadata file: 
 Data Space Used: 6.452GB
 Data Space Total: 21.47GB
 Data Space Available: 15.02GB
 Metadata Space Used: 1.892MB
 Metadata Space Total: 100.7MB
 Metadata Space Available: 98.77MB
 Thin Pool Minimum Free Space: 2.147GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.107-RHEL7 (2015-10-14)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: jvxg46j8bcwa8o5qa96psner9
 Is Manager: true
 ClusterID: et758znsr7ey9izjhlujntroz
 Managers: 4
 Nodes: 4
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Root Rotation In Progress: false
 Node Address: 10.9.4.103
 Manager Addresses:
  10.9.4.103:2377
  10.9.4.117:2377
  10.9.4.122:2377
  10.9.4.212:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 6e23458c129b551d5c9871e5174f6b1b7f6d1170
runc version: 810190ceaa507aa2727d7ae6f4790c76ec150bd2
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-327.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.64GiB
Name: node-10-9-4-103
ID: GC43:XBDK:UUWT:QR3N:S7LB:IHDJ:TETW:2DW5:64UM:G7Q5:GRFJ:OXSO
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 117
 Goroutines: 216
 System Time: 2017-11-15T12:58:17.016590262+01:00
 EventsListeners: 8
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Swarm is setup on 4 manager nodes on VMs with CentOS 7.2 on KVM.

**Comment by thaJeztah** [comment]
DNS entries are removed if the service is not healthy, which may be if (in this case), the service is being rescheduled, failed to rescheduled, and docker keeps trying to schedule until the service becomes healthy.

The `forwarding to ...` messages in the output you provided all seem to coincide with either the service just starting to deploy ("Assigning addresses for endpoint ... ". but no "`addSvcRecords(..)`" yet), or directly after (possibly) the service went unhealthy, and the DNS entries being _removed_ (`deleteSvcRecords(....`).

The last lines of your output show `addSvcRecords(..` followed by `Name To resolve: sa_postgres.` .... `Lookup name sa_postgres. present without IPv6 address`  - which looks to be a "successful" deploy of the service (and name resolution)

**Comment by kostrzewa9ld** [comment]
@thaJeztah 
1. Service `sa_openldap_replica`, the service which is still not resolvable, is reported as `Running`:
```
$ docker service ls --filter name=sa_openldap_replica
ID                  NAME                  MODE                REPLICAS            IMAGE                                       PORTS
ljrvybzqa6b5        sa_openldap_replica   replicated          1/1                 ramencloud/fusiondirectory-openldap:0.3.0   

$ docker service ps sa_openldap_replica
ID                  NAME                        IMAGE                                       NODE                DESIRED STATE       CURRENT STATE         ERROR                              PORTS
xxnkynuwa0xr        sa_openldap_replica.1       ramencloud/fusiondirectory-openldap:0.3.0   node-10-9-4-103     Running             Running 2 days ago                                       
kq869yybtx4b         \_ sa_openldap_replica.1   ramencloud/fusiondirectory-openldap:0.3.0   node-10-9-4-103     Shutdown            Shutdown 2 days ago
```

2. Service `sa_postgres` is resolvable but it took swarm almost 24 hours to stabilize this service.

**Comment by kostrzewa9ld** [comment]
This happened again, with the same Docker version, this time on a freshly deployed set of "healthy" VMs.

Cluster and swarm services are deployed automatically with ansible as part of our CI. This time one of our tests failed already on swarm services deployment because a few of the services were not registered in swarm's DNS although they were reported as running, i.e. passed the health checks and the REPLICAS column showed "1/1".

I can provide debug docker logs from this incident if that can be of any help?

**Comment by kostrzewa9ld** [comment]
I have reproduced this issue with **17.09.1**:
```
Client:
 Version:      17.09.1-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   19e2cf6
 Built:        Thu Dec  7 22:23:40 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.09.1-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   19e2cf6
 Built:        Thu Dec  7 22:25:03 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

I have created 3-node cluster (all managers) and deployed 3 PostgreSQL services with:
```
#!/usr/bin/env bash

export DOCKER_HOST=10.9.4.172
docker network create -d overlay --attachable my_net
docker service create --name postgres1 --health-cmd "pg_isready --host localhost" --health-interval=10s --health-timeout=5s --health-retries=30 --network my_net postgres
docker service create --name postgres2 --health-cmd "pg_isready --host localhost" --health-interval=10s --health-timeout=5s --health-retries=30 --network my_net postgres
docker service create --name postgres3 --health-cmd "pg_isready --host localhost" --health-interval=10s --health-timeout=5s --health-retries=30 --network my_net postgres
```

I have restarted the nodes and although services are marked as running:
```
[xxx@node-10-9-4-172 ~]$ docker service ls
ID                  NAME                MODE                REPLICAS            IMAGE               PORTS
oxt652inewc5        postgres1           replicated          1/1                 postgres:latest     
00hdfek15k1m        postgres2           replicated          1/1                 postgres:latest     
wkz8tl8w5eny        postgres3           replicated          1/1                 postgres:latest  
```

Service `postgres3` was not registered in `my_net` network:
```
[packer@node-10-9-4-172 ~]$ docker network inspect --verbose my_net --format '{{json .Services}}' | jq
{
  "postgres1": {
    "VIP": "10.0.0.2",
    "Ports": [],
    "LocalLBIndex": 257,
    "Tasks": [
      {
        "Name": "postgres1.1.ow2wjloqpnxrskji1fy5ikm2k",
        "EndpointID": "53062ca90fe0452278773c175aad7c4705dd6b4ccac06043504b87a6cccb0d3b",
        "EndpointIP": "10.0.0.3",
        "Info": {
          "Host IP": "10.9.4.217"
        }
      }
    ]
  },
  "postgres2": {
    "VIP": "10.0.0.4",
    "Ports": [],
    "LocalLBIndex": 258,
    "Tasks": [
      {
        "Name": "postgres2.1.y1ubb2r1rp50tll3zmk8v8cgd",
        "EndpointID": "82b1d756725333e77062ebac8e1ed7c8daf3c29568ae1902ffa35565df05064f",
        "EndpointIP": "10.0.0.7",
        "Info": {
          "Host IP": "10.9.4.172"
        }
      }
    ]
  }
}
```

Test with `ping` confirms that:
```
[xxx@node-10-9-4-172 ~]$ docker run -it --rm --network my_net centos ping -c2 -W3 postgres1
PING postgres1 (10.0.0.2) 56(84) bytes of data.
64 bytes from 10.0.0.2 (10.0.0.2): icmp_seq=1 ttl=64 time=0.036 ms
64 bytes from 10.0.0.2 (10.0.0.2): icmp_seq=2 ttl=64 time=0.042 ms

--- postgres1 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1001ms
rtt min/avg/max/mdev = 0.036/0.039/0.042/0.003 ms
[xxx@node-10-9-4-172 ~]$ docker run -it --rm --network my_net centos ping -c2 -W3 postgres2
PING postgres2 (10.0.0.4) 56(84) bytes of data.
64 bytes from 10.0.0.4 (10.0.0.4): icmp_seq=1 ttl=64 time=0.036 ms
64 bytes from 10.0.0.4 (10.0.0.4): icmp_seq=2 ttl=64 time=0.074 ms

--- postgres2 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1001ms
rtt min/avg/max/mdev = 0.036/0.055/0.074/0.019 ms
[xxx@node-10-9-4-172 ~]$ docker run -it --rm --network my_net centos ping -c2 -W3 postgres3
ping: postgres3: Name or service not known
```

**Comment by thaJeztah** [comment]
What does the distribution of those containers look like? Are they all on the same node? Is the missing container on a different node? Logs could be useful indeed.

**Comment by kostrzewa9ld** [comment]
@thaJeztah 
Each container is on a different node:
```
ID                  NAME                IMAGE               NODE                DESIRED STATE       CURRENT STATE            ERROR                              PORTS
ow2wjloqpnxr        postgres1.1         postgres:latest     node-10-9-4-217     Running             Running 19 minutes ago                                      
rbax83pd6yxa         \_ postgres1.1     postgres:latest     node-10-9-4-217     Shutdown            Failed 38 minutes ago    "No such container: postgres1.…"   
ID                  NAME                IMAGE               NODE                DESIRED STATE       CURRENT STATE             ERROR               PORTS
y1ubb2r1rp50        postgres2.1         postgres:latest     node-10-9-4-172     Running             Running 18 minutes ago                        
ntkn7lrh2h1j         \_ postgres2.1     postgres:latest     node-10-9-4-172     Shutdown            Complete 32 minutes ago                       
ID                  NAME                IMAGE               NODE                DESIRED STATE       CURRENT STATE             ERROR               PORTS
d7p1t58t127k        postgres3.1         postgres:latest     node-10-9-4-209     Running             Running 19 minutes ago                        
rulcxdmr1jsy         \_ postgres3.1     postgres:latest     node-10-9-4-209     Shutdown            Complete 19 minutes ago                       
```
Logs from all the nodes are available [here](https://github.com/kostrzewa9ld/issues/tree/master/moby/35508).

**Comment by kostrzewa9ld** [comment]
@thaJeztah 
One more thing worth to mention is that once I got even a more puzzling entry reported by `docker network inspect --verbose`:
```
        "Services": {
            "": {
                "VIP": "<nil>",
                "Ports": [],
                "LocalLBIndex": 0,
                "Tasks": [
                    {
                        "Name": "postgres1.1.w83kc3mc5ine3kl7qn8gyrza3",
                        "EndpointID": "f118eb7bf31e88b4e4668267478796d057b5192bb274cf98789441c5b57e2b97",
                        "EndpointIP": "10.0.0.3",
                        "Info": {
                            "Host IP": "192.168.7.225"
                        }
                    }
                ]
            },
...
```
but that was with 17.06.2. If you believe this deserves a separate issue I can create one as I still have logs from that incident.

**Comment by thaJeztah** [comment]
Quite some errors in the logs for node `10.9.4.209` (the one that looks to be missing in the network inspect); some lines;

```
Dec 20 06:19:14 node-10-9-4-209 dockerd: time="2017-12-20T06:19:14.223972875+01:00" level=error msg="agent: session failed" error="rpc error: code = Internal desc = transport is closing" module="node/agent" node.id=6etidvws1zwfr3huvnfhcj8os
Dec 20 06:19:14 node-10-9-4-209 dockerd: time="2017-12-20T06:19:14.230050123+01:00" level=error msg="agent: session failed" error="rpc error: code = Internal desc = transport: write tcp 10.9.4.209:45670->10.9.4.172:2377: write: broken pipe" module="node/agent" node.id=6etidvws1zwfr3huvnfhcj8os
Dec 20 06:19:14 node-10-9-4-209 dockerd: time="2017-12-20T06:19:14.377933471+01:00" level=error msg="agent: session failed" error="rpc error: code = Unavailable desc = grpc: the connection is unavailable" module="node/agent" node.id=6etidvws1zwfr3huvnfhcj8os
Dec 20 06:19:14 node-10-9-4-209 dockerd: time="2017-12-20T06:19:14.898139273+01:00" level=error msg="agent: session failed" error="rpc error: code = Unavailable desc = grpc: the connection is unavailable" module="node/agent" node.id=6etidvws1zwfr3huvnfhcj8os
Dec 20 06:19:15 node-10-9-4-209 dockerd: time="2017-12-20T06:19:15.023760104+01:00" level=error msg="agent: session failed" error="rpc error: code = Unavailable desc = grpc: the connection is unavailable" module="node/agent" node.id=6etidvws1zwfr3huvnfhcj8os
Dec 20 06:19:17 node-10-9-4-209 dockerd: time="2017-12-20T06:19:17.339118401+01:00" level=error msg="agent: session failed" error="rpc error: code = Unavailable desc = grpc: the connection is unavailable" module="node/agent" node.id=6etidvws1zwfr3huvnfhcj8os
Dec 20 06:19:26 node-10-9-4-209 dockerd: time="2017-12-20T06:19:26.182586920+01:00" level=error msg="failed to create predefined network bridge" error="name conflicts with an existing object" module=node node.id=6etidvws1zwfr3huvnfhcj8os
Dec 20 06:19:26 node-10-9-4-209 dockerd: time="2017-12-20T06:19:26.182656411+01:00" level=error msg="failed to create predefined network host" error="name conflicts with an existing object" module=node node.id=6etidvws1zwfr3huvnfhcj8os
Dec 20 06:19:26 node-10-9-4-209 dockerd: time="2017-12-20T06:19:26.957870005+01:00" level=error msg="agent: session failed" error="session initiation timed out" module="node/agent" node.id=6etidvws1zwfr3huvnfhcj8os
Dec 20 06:22:01 node-10-9-4-209 dockerd: time="2017-12-20T06:22:01.228791089+01:00" level=error msg="agent: session failed" error="session initiation timed out" module="node/agent" node.id=6etidvws1zwfr3huvnfhcj8os
Dec 20 06:22:04 node-10-9-4-209 dockerd: time="2017-12-20T06:22:04.932472704+01:00" level=error msg="agent: session failed" error=EOF module="node/agent" node.id=6etidvws1zwfr3huvnfhcj8os
Dec 20 06:22:06 node-10-9-4-209 dockerd: time="2017-12-20T06:22:06.155176942+01:00" level=error msg="network sa remove failed: unknown network sa id skl5l8hlzytdcevpnqldil0jv" module="node/agent" node.id=6etidvws1zwfr3huvnfhcj8os
Dec 20 06:22:06 node-10-9-4-209 dockerd: time="2017-12-20T06:22:06.155232920+01:00" level=error msg="remove task failed" error="unknown network sa id skl5l8hlzytdcevpnqldil0jv" module="node/agent" node.id=6etidvws1zwfr3huvnfhcj8os task.id=cynpuz9l49494tsyxc6ek9hs8
Dec 20 06:22:06 node-10-9-4-209 dockerd: time="2017-12-20T06:22:06.155182086+01:00" level=error msg="failed removing service binding for 4608edf8364d7ebabf0945377043d7ee82eb6eca8e9c137bebe752bdb1850b2a epRec:{postgres2.1.9swxv8r9tl6vg98947pmkoxj2 postgres2 rr2ya3qg6g8elv9oepsdpi21c 10.0.0.6 10.0.0.7 [] [] [2d2fab1aa4f0]} err:network skl5l8hlzytdcevpnqldil0jv not found"
Dec 20 06:22:09 node-10-9-4-209 dockerd: time="2017-12-20T06:22:09.963698860+01:00" level=error msg="agent: session failed" error="session initiation timed out" module="node/agent" node.id=6etidvws1zwfr3huvnfhcj8os
Dec 20 06:22:15 node-10-9-4-209 dockerd: time="2017-12-20T06:22:15.082484530+01:00" level=error msg="agent: session failed" error="session initiation timed out" module="node/agent" node.id=6etidvws1zwfr3huvnfhcj8os
Dec 20 06:22:16 node-10-9-4-209 dockerd: time="2017-12-20T06:22:16.227590391+01:00" level=error msg="failed to create predefined network bridge" error="name conflicts with an existing object" module=node node.id=6etidvws1zwfr3huvnfhcj8os
Dec 20 06:22:16 node-10-9-4-209 dockerd: time="2017-12-20T06:22:16.227662885+01:00" level=error msg="failed to create predefined network host" error="name conflicts with an existing object" module=node node.id=6etidvws1zwfr3huvnfhcj8os
Dec 20 06:24:56 node-10-9-4-209 dockerd: time="2017-12-20T06:24:56.163488432+01:00" level=error msg="network sa remove failed: network sa not found" module="node/agent" node.id=6etidvws1zwfr3huvnfhcj8os
Dec 20 06:24:56 node-10-9-4-209 dockerd: time="2017-12-20T06:24:56.163563289+01:00" level=error msg="remove task failed" error="network sa not found" module="node/agent" node.id=6etidvws1zwfr3huvnfhcj8os task.id=86628gex1lkbd13hx45wahz6g
Dec 20 06:24:56 node-10-9-4-209 dockerd: time="2017-12-20T06:24:56.163710862+01:00" level=error msg="failed removing service binding for 8208a84bd8d35c3e3ec09cdcfed0378198620d0a387e4532fa3ab42cf0e3d6ac epRec:{postgres2.1.lwfm3cy6xv45tpk097jfbmhao postgres2 rr2ya3qg6g8elv9oepsdpi21c 10.0.0.6 10.0.0.7 [] [] [d20c59fe23d2]} err:network skl5l8hlzytdcevpnqldil0jv not found"
```

And in the `10.9.4.217` node;

```
Dec 20 09:14:16 node-10-9-4-172 dockerd: time="2017-12-20T09:14:16.921855908+01:00" level=error msg="failed removing service binding for 592eb52e796819ac3707b2329e6e7e834c574381220cdfd5edd091085dad21b6 epRec:{nginx.1.i4m3hrr1rzf7l9o128gskvsf1 nginx hugq7mbf9c79k878ldfct5vyo 10.0.0.2 10.0.0.9 [] [] [e8e42bb913cc]} err:network skl5l8hlzytdcevpnqldil0jv not found"
Dec 20 09:19:00 node-10-9-4-172 dockerd: time="2017-12-20T09:19:00.997528344+01:00" level=error msg="Inconsistent driver and libnetwork state for endpoint 7d24b56713e0d99ed13e30fe1f7225c7c12ca7022282ba415741eb5a35a5a0a1"
Dec 20 09:19:09 node-10-9-4-172 dockerd: time="2017-12-20T09:19:09.401144776+01:00" level=error msg="agent: session failed" error="rpc error: code = Canceled desc = context canceled" module="node/agent" node.id=oip26zsv6zv2hnz2sjh918w9j
Dec 20 09:19:09 node-10-9-4-172 dockerd: time="2017-12-20T09:19:09.432795875+01:00" level=error msg="agent: session failed" error="rpc error: code = Unavailable desc = grpc: the connection is unavailable" module="node/agent" node.id=oip26zsv6zv2hnz2sjh918w9j
Dec 20 09:19:09 node-10-9-4-172 dockerd: time="2017-12-20T09:19:09.633098699+01:00" level=error msg="agent: session failed" error="rpc error: code = Unavailable desc = grpc: the connection is unavailable" module="node/agent" node.id=oip26zsv6zv2hnz2sjh918w9j
```

"Inconsistent driver and libnetwork state for endpoint" is coming from https://github.com/docker/libnetwork/blob/17fccdeb660b4bb6ced59e26ccfab75888d1e35f/agent.go#L465-L473

ping @ddebroy @fcrisciani - anything of interest in the logs? (all logs can be found here; https://github.com/kostrzewa9ld/issues/tree/master/moby/35508)

**Comment by kondziolka9ld** [comment]
I came across similar issue using following setup:

Docker: `docker-ce-18.03.1`
OS: `Red Hat 7.2`
Linux version: `3.10.0-514.el7.x86_64`
Number of nodes: `3` (each of them is worker and manager)

One of services (**myservice**) was not registered in internal DNS at all, please look at below logs. Nevertheless, **myservice** service was properly started (according to below logs). As a result, another services in my cluster weren't able to resolve name **myservice**.
```
Oct 19 00:31:26 node-10-9-4-36 dockerd: time="2018-10-19T00:31:26.565071225Z" level=debug msg=assigned module=node/agent node.id=sy7x5zi0ep5t4if5ahteeg9vz task.desiredstate=RUNNING task.id=wvg5xv91h6l1lz6ls28n0qyv9
Oct 19 00:31:26 node-10-9-4-36 dockerd: time="2018-10-19T00:31:26.565441263Z" level=debug msg="state changed" module=node/agent node.id=sy7x5zi0ep5t4if5ahteeg9vz service.id=cicdg4s76t1ie92xtjnbiqxbi state.desired=RUNNING state.transition="ASSIGNED->ACCEPTED" task.id=wvg5xv91h6l1lz6ls28n0qyv9
Oct 19 00:31:26 node-10-9-4-36 dockerd: time="2018-10-19T00:31:26.565565395Z" level=debug msg="(*Agent).UpdateTaskStatus" module=node/agent node.id=sy7x5zi0ep5t4if5ahteeg9vz task.id=wvg5xv91h6l1lz6ls28n0qyv9
Oct 19 00:31:26 node-10-9-4-36 dockerd: time="2018-10-19T00:31:26.565665984Z" level=debug msg="state changed" module=node/agent/taskmanager node.id=sy7x5zi0ep5t4if5ahteeg9vz service.id=cicdg4s76t1ie92xtjnbiqxbi state.desired=RUNNING state.transition="ASSIGNED->ACCEPTED" task.id=wvg5xv91h6l1lz6ls28n0qyv9
Oct 19 00:31:26 node-10-9-4-36 dockerd: time="2018-10-19T00:31:26.567918660Z" level=debug msg="task status reported" module=node/agent node.id=sy7x5zi0ep5t4if5ahteeg9vz
Oct 19 00:31:26 node-10-9-4-36 dockerd: time="2018-10-19T00:31:26.567952065Z" level=debug msg="(*Agent).UpdateTaskStatus" module=node/agent node.id=sy7x5zi0ep5t4if5ahteeg9vz task.id=wvg5xv91h6l1lz6ls28n0qyv9
Oct 19 00:31:26 node-10-9-4-36 dockerd: time="2018-10-19T00:31:26.568591854Z" level=debug msg="task status reported" module=node/agent node.id=sy7x5zi0ep5t4if5ahteeg9vz
Oct 19 00:31:26 node-10-9-4-36 dockerd: time="2018-10-19T00:31:26.570085748Z" level=debug msg="state changed" module=node/agent/taskmanager node.id=sy7x5zi0ep5t4if5ahteeg9vz service.id=cicdg4s76t1ie92xtjnbiqxbi state.desired=RUNNING state.transition="ACCEPTED->PREPARING" task.id=wvg5xv91h6l1lz6ls28n0qyv9
Oct 19 00:31:26 node-10-9-4-36 dockerd: time="2018-10-19T00:31:26.570507519Z" level=debug msg="(*Agent).UpdateTaskStatus" module=node/agent node.id=sy7x5zi0ep5t4if5ahteeg9vz task.id=wvg5xv91h6l1lz6ls28n0qyv9
Oct 19 00:31:28 node-10-9-4-36 dockerd: time="2018-10-19T00:31:28.719803021Z" level=debug msg="Creating service config in agent for t = &Task{ID:wvg5xv91h6l1lz6ls28n0qyv9,Meta:Meta{Version:Version{Index:133,},CreatedAt:2018-10-19T00:31:26.396335652Z,UpdatedAt:2018-10-19T00:31:26.459320778Z,},Spec:TaskSpec{Runtime:&TaskSpec_Container{Container:&ContainerSpec{Image:mycompany/myservice:1.1.7-43-g92e1a4b,Labels:map[string]string{com.docker.stack.namespace: mystack,},Command:[],Args:[],Env:[],Dir:,User:,Mounts:[],StopGracePeriod:(duration: nil google_protobuf1.Duration),PullOptions:nil,Groups:[],Secrets:[],TTY:false,Hostname:,DNSConfig:nil,Healthcheck:&HealthConfig{Test:[CMD-SHELL curl -f http://localhost:8998],Interval:10s,Timeout:5s,Retries:30,StartPeriod:0s,},Hosts:[],OpenStdin:false,ReadOnly:false,StopSignal:,Configs:[&ConfigReference{ConfigID:q8acwvpxejd5i9iami8heq15h,ConfigName:s
Oct 19 00:31:28 node-10-9-4-36 dockerd: time="2018-10-19T00:31:28.719939372Z" level=debug msg="state changed" module=node/agent/taskmanager node.id=sy7x5zi0ep5t4if5ahteeg9vz service.id=cicdg4s76t1ie92xtjnbiqxbi state.desired=RUNNING state.transition="PREPARING->READY" task.id=wvg5xv91h6l1lz6ls28n0qyv9
Oct 19 00:31:28 node-10-9-4-36 dockerd: time="2018-10-19T00:31:28.720362264Z" level=debug msg="(*Agent).UpdateTaskStatus" module=node/agent node.id=sy7x5zi0ep5t4if5ahteeg9vz task.id=wvg5xv91h6l1lz6ls28n0qyv9
Oct 19 00:31:28 node-10-9-4-36 dockerd: time="2018-10-19T00:31:28.721821989Z" level=debug msg="state changed" module=node/agent/taskmanager node.id=sy7x5zi0ep5t4if5ahteeg9vz service.id=cicdg4s76t1ie92xtjnbiqxbi state.desired=RUNNING state.transition="READY->STARTING" task.id=wvg5xv91h6l1lz6ls28n0qyv9
Oct 19 00:31:28 node-10-9-4-36 dockerd: time="2018-10-19T00:31:28.929701911Z" level=debug msg="Assigning addresses for endpoint sa_myservice.1.wvg5xv91h6l1lz6ls28n0qyv9's interface on network ingress"
Oct 19 00:31:28 node-10-9-4-36 dockerd: time="2018-10-19T00:31:28.931559862Z" level=debug msg="Assigning addresses for endpoint sa_myservice.1.wvg5xv91h6l1lz6ls28n0qyv9's interface on network ingress"
Oct 19 00:31:28 node-10-9-4-36 dockerd: time="2018-10-19T00:31:28.962189913Z" level=debug msg="Assigning addresses for endpoint sa_myservice.1.wvg5xv91h6l1lz6ls28n0qyv9's interface on network mynetwork"
Oct 19 00:31:28 node-10-9-4-36 dockerd: time="2018-10-19T00:31:28.963671126Z" level=debug msg="Assigning addresses for endpoint sa_myservice.1.wvg5xv91h6l1lz6ls28n0qyv9's interface on network mynetwork"
Oct 19 00:31:29 node-10-9-4-36 dockerd: time="2018-10-19T00:31:29.456862977Z" level=debug msg="waiting on events" module=node/agent/taskmanager node.id=sy7x5zi0ep5t4if5ahteeg9vz service.id=cicdg4s76t1ie92xtjnbiqxbi task.id=wvg5xv91h6l1lz6ls28n0qyv9
Oct 19 00:31:33 node-10-9-4-36 dockerd: time="2018-10-19T00:31:33.278798856Z" level=debug msg="Name To resolve: service_myservice."
Oct 19 00:31:33 node-10-9-4-36 dockerd: time="2018-10-19T00:31:33.278925160Z" level=debug msg="[resolver] query service_myservice. (A) from 172.18.0.6:54066, forwarding to udp:10.9.4.4"
Oct 19 00:31:33 node-10-9-4-36 dockerd: time="2018-10-19T00:31:33.377310588Z" level=debug msg="[resolver] external DNS udp:10.9.4.4 did not return any A records for \"service_myservice.\""
Oct 19 00:31:33 node-10-9-4-36 dockerd: time="2018-10-19T00:31:33.377660647Z" level=debug msg="Name To resolve: service_myservice.bdcluster.."
Oct 19 00:31:33 node-10-9-4-36 dockerd: time="2018-10-19T00:31:33.377772572Z" level=debug msg="[resolver] query service_myservice.bdcluster.. (A) from 172.18.0.6:48280, forwarding to udp:10.9.4.4"
Oct 19 00:31:33 node-10-9-4-36 dockerd: time="2018-10-19T00:31:33.383109406Z" level=debug msg="[resolver] external DNS udp:10.9.4.4 did not return any A records for \"service_myservice.bdcluster..\""
Oct 19 00:31:38 node-10-9-4-36 dockerd: time="2018-10-19T00:31:38.618863351Z" level=debug msg=assigned module=node/agent node.id=sy7x5zi0ep5t4if5ahteeg9vz task.desiredstate=RUNNING task.id=wvg5xv91h6l1lz6ls28n0qyv9
Oct 19 00:31:39 node-10-9-4-36 dockerd: time="2018-10-19T00:31:39.149057946Z" level=debug msg="(*Agent).UpdateTaskStatus" module=node/agent node.id=sy7x5zi0ep5t4if5ahteeg9vz task.id=wvg5xv91h6l1lz6ls28n0qyv9
Oct 19 00:31:40 node-10-9-4-36 dockerd: time="2018-10-19T00:31:40.151156924Z" level=debug msg="state changed" module=node/agent/taskmanager node.id=sy7x5zi0ep5t4if5ahteeg9vz service.id=cicdg4s76t1ie92xtjnbiqxbi state.desired=RUNNING state.transition="STARTING->RUNNING" task.id=wvg5xv91h6l1lz6ls28n0qyv9
Oct 19 00:31:40 node-10-9-4-36 dockerd: time="2018-10-19T00:31:40.151458703Z" level=debug msg="(*Agent).UpdateTaskStatus" module=node/agent node.id=sy7x5zi0ep5t4if5ahteeg9vz task.id=wvg5xv91h6l1lz6ls28n0qyv9
Oct 19 00:31:40 node-10-9-4-36 dockerd: time="2018-10-19T00:31:40.152968487Z" level=debug msg="waiting on events" module=node/agent/taskmanager node.id=sy7x5zi0ep5t4if5ahteeg9vz service.id=cicdg4s76t1ie92xtjnbiqxbi task.id=wvg5xv91h6l1lz6ls28n0qyv9
Oct 19 00:31:58 node-10-9-4-36 dockerd: time="2018-10-19T00:31:58.455764837Z" level=debug msg=assigned module=node/agent node.id=sy7x5zi0ep5t4if5ahteeg9vz task.desiredstate=RUNNING task.id=wvg5xv91h6l1lz6ls28n0qyv9
```

There is no another logs of `dockerd` for **myservice** and corresponding task (`wvg5xv91h6l1lz6ls28n0qyv9`), in particular there is neither log related to `addServiceBinding` nor `handleEpTableEvent`. Obviously, `docker network inspect mynetwork` doesn't show **myservice** in Services section. Strangely enough, it shows container corresponding to **myservice** among Containers:

```docker network inspect mynetwork
     {
        "Name": "mynetwork",
        "Id": "ploejk596pd892d0hvvyad7o4",
        [..]
        "Containers": {
            [...]
            "bb03c6af6d89eedded9622751af335099da6876a9844d70bd9be5b6de15ff48a": {
                "Name": "sa_myservice.1.wvg5xv91h6l1lz6ls28n0qyv9",
                "EndpointID": "cae370e7ff9b3d89265aa48784aada1a348d1597f3c53709d8307fb480eeaca6",
                "MacAddress": "02:42:0a:00:00:06",
                "IPv4Address": "10.0.0.6/24",
                "IPv6Address": ""
            }
        },
```
It is difficult for me to give exact reproducible scenario due to non-deterministic nature of this issue.

**Comment by thaJeztah** [comment]
I'm a bit confused by the logs you provided and your example description; a service `sa_myservice` is started, but the logs show you're trying to connect to a service `service_myservice`, probably attached to a `bdcluster` network.

Your case is likely different from the original issue reported here, which had a lot of errors in the logs, so if you suspect there's a bug at hand, it may be better to open an issue with details (at least `docker version` and `docker info` output, and if possible a minimal case to reproduce)

**Comment by kondziolka9ld** [comment]
> I'm a bit confused by the logs you provided and your example description; a service sa_myservice is started, but the logs show you're trying to connect to a service service_myservice, probably attached to a bdcluster network

When it comes to connecting to `service_myservice.bdcluster` service it is not related to this issue - some client tries to resolve DNS name, however it doesn't matter for this issue. 

The problem is in another client:
`Could not establish connection to [myservice:8998] due to java.net.UnknownHostException: myservice`

> Your case is likely different from the original issue reported here, which had a lot of errors in the logs

I believe that root cause is very similar or even the same - in both cases there is no information about registering some services in internal Swarm DNS (while as for another services such information is present). Please note that it occurs occasionally in our CI environment so it is non-deterministic. At this moment, I have no access to this cluster to provide you `docker info` and `docker version`. However, information provided in my previous comment was reliable.

**Comment by kevinbrink** [comment]
Has there been any movement on this?

I encountered what I believe is the same problem earlier this morning in our production environment. 

## Environment:

|Docker Version|17.12.0-ce-dind|
|---|---|
|OS|Minimalistic linuxkit-based OS|
|Kernel|4.15.5|
|Manager Nodes|3|
|Worker nodes|6|

## Scenario

At some point, our system failed for a reason yet to be determined (presumably unrelated to this issue). Part of the troubleshooting involved restarting the VM's.

When the VM's came back up, the services began to start up, but one of them failed to register with swarm's DNS, despite the fact that the container was running and `docker service ls` showed it having `1/1` replicas.

As a test, I spun up a new container attached to the stack's network and tried to `ping` various services inside of the stack. Other running services responded fine, but pinging this unregistered service resulted in the `Name or service not known` error.

Performing a blind `docker service update --force <service-name>` to restart it "fixed" the problem, as it started a new task instance, but I am obviously a little worried about this randomly happening again. The new task instance happened to be spun up on the same node, so I don't think this issue has anything to do with which node the task is running. 

## Additional information:

I `inspect`ed the two service containers, the one that was unregistered / orphaned and the one that started working when I restarted the service. What I believe to be the interesting differences are below:

 * `+` lines are from the _currently running, working service container_
 * `-` lines are from the old service container that was not accessible from the network. 
```diff
[
    {
        ...
        "NetworkSettings": {
            ...
-           "Ports": {},
+           "Ports": {
+               "15671/tcp": null,
+               "15672/tcp": null,
+               "25672/tcp": null,
+               "4369/tcp": null,
+               "5671/tcp": null,
+               "5672/tcp": null
+           },
            "Networks": {
                "app_network": {
-                   "EndpointID": "",
+                   "EndpointID": "093bacfba302ac025a966baae79c27b10dc3440adf0b7336f9b4c3140bfca817",
                    "Gateway": "",
-                   "IPAddress": "",
+                   "IPAddress": "10.0.0.133",
-                   "IPPrefixLen": 0,
+                   "IPPrefixLen": 24,
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
-                   "MacAddress": "",
+                   "MacAddress": "02:42:0a:00:00:85",
                    "DriverOpts": null
                }
            }
        }
    }
]
```

I also did some digging in the docker daemon logs for that node. I can post more logs if it might be helpful, but the following lines are from when the successful container got started on the node. These lines (e.g., `addServiceBinding` and `addEndpointNameResolution`) were **not** present in the logs from when the unregistered task instance container got started; it's as if docker skipped these steps when spinning up the first container that ended up staying unregistered:

```
addServiceBinding from addServiceInfoToCluster START for app_my_service 093bacfba302ac025a966baae79c27b10dc3440adf0b7336f9b4c3140bfca817 p:0xc42036ab80 nid:fxx4vl567b0vdkhidzfo4c5fo skey:{nx43qcnldjqu1byyzokr390s2 }"
addEndpointNameResolution 093bacfba302ac025a966baae79c27b10dc3440adf0b7336f9b4c3140bfca817 app_my_service add_service:true sAliases:[my_service] tAliases:[89f35c1c52fd]"
093bacfba302ac025a966baae79c27b10dc3440adf0b7336f9b4c3140bfca817 (fxx4vl5).addSvcRecords(app_my_service.1.a50visossfiwvugumct6cjw6z, 10.0.0.133, <nil>, true) addServiceBinding sid:093bacfba302ac025a966baae79c27b10dc3440adf0b7336f9b4c3140bfca817"
093bacfba302ac025a966baae79c27b10dc3440adf0b7336f9b4c3140bfca817 (fxx4vl5).addSvcRecords(89f35c1c52fd, 10.0.0.133, <nil>, true) addServiceBinding sid:093bacfba302ac025a966baae79c27b10dc3440adf0b7336f9b4c3140bfca817"
093bacfba302ac025a966baae79c27b10dc3440adf0b7336f9b4c3140bfca817 (fxx4vl5).addSvcRecords(tasks.app_my_service, 10.0.0.133, <nil>, false) addServiceBinding sid:nx43qcnldjqu1byyzokr390s2"
093bacfba302ac025a966baae79c27b10dc3440adf0b7336f9b4c3140bfca817 (fxx4vl5).addSvcRecords(tasks.my_service, 10.0.0.133, <nil>, false) addServiceBinding sid:nx43qcnldjqu1byyzokr390s2"
093bacfba302ac025a966baae79c27b10dc3440adf0b7336f9b4c3140bfca817 (fxx4vl5).addSvcRecords(app_my_service, 10.0.0.59, <nil>, false) addServiceBinding sid:nx43qcnldjqu1byyzokr390s2"
093bacfba302ac025a966baae79c27b10dc3440adf0b7336f9b4c3140bfca817 (fxx4vl5).addSvcRecords(my_service, 10.0.0.59, <nil>, false) addServiceBinding sid:nx43qcnldjqu1byyzokr390s2"
```

When the unregistered instance was running, logs like this appeared, presumably when other services were trying to query for this service:

```
time="2018-11-22T13:29:53.484280800Z" level=debug msg="Name To resolve: my_service."
time="2018-11-22T13:29:53.484447667Z" level=debug msg="[resolver] query my_service. (AAAA) from 172.18.0.6:41924, forwarding to udp:8.8.8.8"
time="2018-11-22T13:29:53.484502211Z" level=debug msg="[resolver] query my_service. (A) from 172.18.0.6:59840, forwarding to udp:8.8.8.8"
```

However, when the service was functioning properly, there were logs like this:

```
time="2018-11-22T13:32:39.601879978Z" level=debug msg="Name To resolve: my_service."
time="2018-11-22T13:32:39.601944729Z" level=debug msg="[resolver] lookup for my_service.: IP [10.0.0.59]"
time="2018-11-22T13:32:39.601891442Z" level=debug msg="[resolver] lookup name my_service. present without IPv6 address"
```

Let me know if I can be of any further assistance with logs, and let me know if there's any upcoming fixes for this, or any mitigation strategies.

I must say that, to me, has particular issue been the last straw in a line of frustrations with Docker swarm. We are not in a position to simply switch infrastructures, but I have become convinced that choosing Docker swarm to run our services on was a mistake. Having services that can run for hours without being accessible by other services seems like a fundamental breakage in what docker swarm promises.

**Comment by thaJeztah** [comment]
@kevinbrink do any of those services use healthchecks? That could be an explanation (of a task is not marked "healthy", no DNS entries are added for the service. Not sure if that's an explanation in your case though.

Things I _do_ notice is that you're running Docker 17.12, which reached end of live in March 2018. If there's specific reasons not to update, make sure you're running the last patch release of 17.12 ([17.12.1](https://github.com/docker/docker-ce/releases/tag/v17.12.1-ce))

**Comment by kevinbrink** [comment]
@thaJeztah : Thanks for the response.

The comparison I was drawing was between two instances of the same exact service. That service does not have any sort of health check, so I don't think that's a factor. 

Our infrastructure is such that updating docker can't happen as often as we'd like. We are preparing to update to `18.06.1`, but I don't have a ton of confidence this particular issue has been mitigated by that upgrade.

**Comment by tonygolden** [comment]
@thaJeztah
Maybe, I have found something useful. 

1. When new leader is selected, TaskManagers of the workers would receive the task info from the new leader. And this may interrupt the task flow in the TaskManager.

2. When the task with health-check enabled is in StartingState, it would be:
2-1 start the container
2-1 when the container is healthy, register the DNS entry.

Between 2-1 and 2-2 it would take seconds, at this time 1 comes and then the task workflow is interrupted. The DNS entry is missing now until I use "docker service update" to recover it.

**Comment by thaJeztah** [comment]
@dperny @mavenugo could you have a look?

**Comment by tonygolden** [comment]
@dperny @mavenugo @thaJeztah 
Is there any progress on this issue?
